README.md
# Automated Risk Register Review and Compare (AuReCom)

This project is created as a tool for my monthly project risk reviews. Currently attached to a construction program where I need to manage 10 risk registers, in other words, 10 projects. It is easier to create a script that will do it for you although learning python (even coding) as you code is quite a challenge.

As this is my first program, there are many inconsistencies as i was exploring or experimenting new ways to execute what i wanted.

The first part of AuReCom is the risk register review while the second part will compare the risk register to previous month and identify what has changed and what has been accomplished.

Initially, the idea came after i learned using openpyxl and straight started this project. Then stumbled upon using pandas for the second part. The project started with manually assign specific cells in excel for checking and review then transition to automatically select which cell to look at, identify the header row and etc.

The script will initially request for an excel file downloaded from our Risk Management System in a specific format as first input. Then it'll start to flag the risk register for errors and cells or items that require attention (risks triggering soon, risks expired and expiring soon, mitigation actions due and overdue, etc). It'll then write a status update in a new cell next to the end column of first row of the current risk register by giving the number of risks, risk status, and actions status. 

New excel file created and saved.

Then it'll request for another excel file for comparison, this is usually previous month's risk register that i am reviewing. The reason for comparing to last month's risk register is that i can track what have the projects accomplished in their risk management activities. Changes will be captured in a new sheet and distinguished with "old --<<>>-- new" format. At the same time changes that i care about (i.e changes to mitigation actions) are interpreted and added to a new column in this new sheet.

During this stage, the script will also identify which risks are new and which are closed or retired at the time of review in new sheets.

Finally, all of the changes that i care about will be shown in the same cell as the status update.

Then saved on the new excel file created.

Additionally the script is converted to .exe so my other teammates can make a good use out of it.


changelog.txt
#  Changelog
#
#  2.1 user input update and better action & control cross referencing
#  2.1 added tkinter import tk to hide tk window when running the program
#  2.2 delete comment import warnings
#  2.2 update comments to be more descriptive
#  2.2 added simple instructions to user
#  2.3 added total number of active actions so that we can find the ratio of overdue and active actions
#  2.3.1 issue encountered - empty line and spaces on actions/controls will be redFilled. solution, remove them at cxreflist using .replace()
#  2.3.1 use regular expression (regex) for current controls and treatment actions for cross referencing c1,c2,c3...etc
#  2.3.1 added new list for action and current controls named valid[] for words that are valid so it wont get redfilled
#  2.3.2 added stats on active & completed actions
#  2.3.2 added exception handling for zero division error when calculating percentage of overdue actions
#  2.3.3 added a for loop for completed actions so we can append to list instead of skip # not using this #
#  2.3.3 the enumerate for loop for treatment actions taken out of its if statement and given priority to upper line of code. this is to prevent skipping if the actions are all completed and not being counted.
#  2.3.4 problem comparing risk registers when get.duplicates returned blank list. need to rectify...
#  2.3.4 added if statement checking whether changed_risks is empty. since this is a panel, it wont give a dataframe if there are no changes. so quick fix to assign an empty dataframe i.e change_new (this is empty because this is one of our panel df)
#  2.4 file_path is changed to get the file name (without extension) and added 'comments.xlsx' string to the end. this will ensure that the file saved will be in the same name.
#  2.5 skip Description, Reviewed By, Last Review, Next Review Date, Last Review Note columns
#  2.5 added code to red flag active treatment actions but rating the same
#  2.5.1 Description, Reviewed By, Last Review, Next Review Date, Last Review Note columns to be reviewed if the risk is high
#  2.6 Add count proposed actions
#  2.6.1 change actions count to the Action Status column
#  2.6.1 clean currate curratelist duplication
#  2.6.1 add !!! to proposed actions and red flag
#  2.7 skip redflag if action is abandon but no owner
#  2.8 skip blank treatment action if current rating is low
#  2.9 import itertools zip_longest - this is to merge 2 list element wise based on longest list.
#  2.9 compare difference on what's changed for action status with another risk register.
#  2.9.1 to run action tracker only on 'Changed' worksheet.
#  2.9.1 add new column 'Actions Tracker' to 'Changed' worksheet
#  2.9.2 changed from 'append' list to 'extend' for the action tracker
#  2.9.2 modify 'Actions Tracker' column width to be wider in the .csv file. This updates other worksheets as well
#  2.9.3 add total 'Actions Tracker'
#  2.9.4 modify total 'Actions Tracker' to show in 'Comments' sheet
#  3.0 touch up and clean up. Then combine the actions tracker into 1 cell.
#  3.1 added extra comments on the 'Comments' row on top of register
#  3.1 marked '!!!' to undefined action owners for Active and Completed actions
#  3.1.1 red flag both current and target rating when actions are all completed if not assessed
#  3.1.1 tarrate value not being updated, thus move the line code to the top when columnno = description etc.
#  3.2 blueFill colour to current and target rating if actions are completed
#  3.2 tanFill colour to current and target rating if there is still active action but rating the same
#  3.2 added legend at top on what the colours mean
#  3.3 added rating tracker
#  3.3.1 combine rating and action tracker. even if in the same row.
#  3.4 action and rating tracker now detect riskID. first, compare row which RiskID in which row
#  3.4.1 display tracker at comment sheet. now at the column u want. now needs formatting and column title.
#  3.4.2 format tracker at comment sheet. only if there is changes at current rating or action status
#  3.4.2 delete 'Actions Tracker' column in 'changed' worksheet. delete corresponding column width in columnwidth.csv file
#  3.5 add actions tracker for actions made active
#  3.5.1 green flag for the changes in 'comments' worksheet
#  3.5.1 change variable names so easy to understand. e.g final count, all actionsID, cell_final. etc
#  3.5.2 define action_row and rating_row prior to checking cell. if there isnt any risk rating change or action change, those 2 variable wont be defined and give error
#  3.5.3 re-enable 'Actions Tracker' column for sanity check...cell_changed created to reflect cell at "Changed" sheet
#  3.5.4 re-arrange variables closer to where it'll be called
#  3.5.5 firstrow...modify variable name to reflect whats being used, instead of header_df +1 +2 +3 etc. add few variables for firstrows: header_df, header for excel, sre comments row
#  3.5.6 change the old df on top to new df.
#  3.5.7 remove looping for column width. this can be done in looping for red flag below it.
#  3.5.7 remove define sheets i.e sheetcomments, sheetchanged, etc. this can be done in looping below them. sheetchanged is already define as sheet on top.
#  3.6 add risk counter plakkk using pandas 
#  3.6.1 add final_count variable. senang sikit. then finalize to final_summary
#  3.7 add risk changes counter plakkk. new variable - no_rating_increased / decreased
#  3.8 add new and closed risk counter plakkk
#  3.8 rename few variables for easy understanding
#  4.0 modify code for new column ARM License Holder: only change .column_dimensions['column letters'].width = 50 & .column_dimensions['column letters'].hidden = False
#  4.0 modify columnwidth.csv to adjust column accordingly
#  4.0 add signature
#  4.1 get_column_letter based on no...modify columnlist_ori so we have original state of columns

compare3.4.1.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#  comp2.py
#  
#  Copyright 2017 yus22159 <yus22159@RP0005577591>
#  
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#  
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.
#  
#  

import pandas as pd
import collections
import csv
from openpyxl import load_workbook
from openpyxl.styles import Color, PatternFill, Font, Border, Alignment, Side

# The previous method of using the dropna gives error when both files contain same nan column. this method does not use dropna but select the columns we want instead when saving.

# Define the diff function to show the changes in each field
def report_diff(x):
    if x[0] != x[1]:# and (x[0] is not None or x[0] != 'nan'):
        return ('{}\n--<<>>--\n{}'.format(*x))
    else:
        return x[0]
	#return x[0] if x[0] == x[1] else '{}\n--<<>>--\n{}'.format(*x)

alignment = Alignment(horizontal='general',
                      vertical='top',
                      text_rotation=0,
                      wrap_text=True,
                      shrink_to_fit=False,
                      indent=0)

border = Border(left=Side(border_style='thin',
                          color='B2B2B2'),
                right=Side(border_style='thin',
                           color='B2B2B2'),
                top=Side(border_style='thin',
                         color='B2B2B2'),
                bottom=Side(border_style='thin',
                            color='B2B2B2'))

redFill = PatternFill(start_color='FFC7CE',
                      end_color='FFC7CE',
                      fill_type='solid')

greyFill = PatternFill(start_color='DCDCDC',
                      end_color='DCDCDC',
                      fill_type='solid')

# disable SettingWithCopyWarning 
pd.options.mode.chained_assignment = None  # default='warn'

#getting the first row and all column name
#read excel with no header to get accurate row number and fill nan with 'N/A' for easy tracking
old = pd.read_excel('03.xlsx',header=None).fillna('N/A')
#iterate rows into tuples then into list
cycle=list(old.itertuples(index=False))
#empty list for our column name
columnlist=[]

firstrow=0
while firstrow < len(cycle):
  for colu in cycle[firstrow]:
    columnlist.append(colu)
  counter=collections.Counter(columnlist)
  if counter['N/A'] < 5:
    break
  firstrow+=1
  columnlist=[]

#final column name
columnlist = [x for x in columnlist if x != 'N/A']

# Read in the two files but call the data old and new and create columns to track
old = pd.read_excel('old.xlsx', header=firstrow)#.dropna(axis=1,how='all') if i dont dropna now, columns positions are still in placed
new = pd.read_excel('03.xlsx', header=firstrow)#.dropna(axis=1,how='all') 
old['version'] = "old"
new['version'] = "new"

# Set the columns you want to check to remove duplicate

# Set the order of the columns that'll be sorted when saving later
# Since the columns order is automatically retrieved, the lists are not needed

# empty list to get column values
colwidth = {}

#Join all the data together and ignore indexes so it all gets added
full_set = pd.concat([old,new],ignore_index=True).fillna('') #dropna here and columns order is still intact. not using anymore...

# Let's see what changes in the main columns we care about - subset is crucial if not it won't remove duplicates'
changes = full_set.drop_duplicates(subset=columnlist,keep='last')

#We want to know where the duplicate RiskIDs are, that means there have been changes. this return only the duplicated RiskID
dupe_risks = changes.set_index("ID").index.get_duplicates()

#Get all the duplicate rows. Subset dataframe
dupes = changes[changes["ID"].isin(dupe_risks)]

#Pull out the old and new data into separate dataframes
change_new = dupes[(dupes["version"] == "new")]
change_old = dupes[(dupes["version"] == "old")]

#Drop the temp columns - we don't need them now
change_new = change_new.drop(['version'], axis=1)
change_old = change_old.drop(['version'], axis=1)

#Index on the RiskID
change_new.set_index('ID',inplace=True)
change_old.set_index('ID',inplace=True)

#Now we can diff because we have two data sets of the same size with the same index
diff_panel = pd.Panel(dict(df1=change_old,df2=change_new))
changed_risks = diff_panel.apply(report_diff, axis=0)

#Diff'ing is done, we need to get a list of removed items

#Flag all duplicated RiskIDs
changes['duplicate']=changes["ID"].isin(dupe_risks)

#Identify non-duplicated items that are in the old version and did not show in the new version
closed_risks = changes[(changes["duplicate"] == False) & (changes["version"] == "old")]

#Re-Index Closed Risks
closed_risks.set_index('ID',inplace=True)

# We have the old and diff, we need to figure out which ones are new

#Drop duplicates but keep the first item instead of the last
new_risk_set = full_set.drop_duplicates(subset=columnlist,keep='first')

#Identify dupes in this new dataframe
new_risk_set['duplicate']=new_risk_set["ID"].isin(dupe_risks)

#Identify added risks
added_risks = new_risk_set[(new_risk_set["duplicate"] == False) & (new_risk_set["version"] == "new")]

#Re-Index added risks
added_risks.set_index('ID',inplace=True)

#Save the changes to excel but only include the columns we care about
book=load_workbook("03-comments.xlsx")
writer = pd.ExcelWriter("03-comments.xlsx",engine='openpyxl')
writer.book=book

#write on top of active sheet - no overwrite - no adding new sheets
writer.sheets=dict((ws.title,ws) for ws in book.worksheets)

#print(changed_risks.index.name) #to remove the index column from the columnlist
columnlist.remove(changed_risks.index.name) 
#save to an existing file in a new sheet and select the columns we want
changed_risks.to_excel(writer,"Changed", index=True, columns=columnlist)
closed_risks.to_excel(writer,"Closed",index=True, columns=columnlist)
added_risks.to_excel(writer,"New",index=True, columns=columnlist)

#adjusting column width
sheetchanged = book.get_sheet_by_name('Changed')
sheetclosed = book.get_sheet_by_name('Closed')
sheetnew = book.get_sheet_by_name('New')

#get column values from output file into colwidth={} dictionary
for key, val in csv.reader(open("columnwidth.csv")):
    colwidth[key] = val

#apply column values to new sheets
for col_letters in colwidth:
  sheetchanged.column_dimensions[col_letters].width = colwidth[col_letters]
  sheetclosed.column_dimensions[col_letters].width = colwidth[col_letters]
  sheetnew.column_dimensions[col_letters].width = colwidth[col_letters]

#apply format to new sheets
for shet in book.worksheets:
  if shet == book.worksheets[0]:
    continue
  else:
    for rowNo in range(1, shet.max_row + 1):
      for columnNo in range(1, shet.max_column + 1):
        cell = shet.cell(row=rowNo, column=columnNo)
        cell.value = str(cell.value)
        cell.alignment = alignment
        cell.border = border
        if rowNo==1:
          cell.fill = greyFill
        #elif cell.value=="nan\n--<<>>--\nnan" or cell.value=="NaT\n--<<>>--\nNaT":
          #cell.value=None
        elif cell.value.find("<<>>") != -1 and cell.value is not None:
          cell.fill=redFill

writer.save()


riskreview-v4.9.py
# auto review risk register
import openpyxl
from openpyxl.styles import Color, PatternFill, Font, Border, Alignment, Side
import time
from datetime import datetime, date
from tkinter.filedialog import askopenfilename

today = date.today()  # getting the date today.
# todaysdate=time.strftime("%d %b %y")
# print(todaysdate)

filename=askopenfilename()
if filename is '':
	print('No file selected')
	quit()

wb = openpyxl.load_workbook(filename)
sheet = wb.get_sheet_by_name('03. RAPID Risk Register')

redFill = PatternFill(start_color='FFC7CE',
                      end_color='FFC7CE',
                      fill_type='solid')
yellowFill = PatternFill(start_color='FFFFCC', end_color='FFFFCC',
                         fill_type='solid')
italic = Font(size=10, italic=True)
alignment = Alignment(horizontal='general',
                      vertical='top',
                      text_rotation=0,
                      wrap_text=True,
                      shrink_to_fit=False,
                      indent=0)
border = Border(left=Side(border_style='thin',
                          color='B2B2B2'),
                right=Side(border_style='thin',
                           color='B2B2B2'),
                top=Side(border_style='thin',
                         color='B2B2B2'),
                bottom=Side(border_style='thin',
                            color='B2B2B2'))

def splitline (abc):
	return abc.split('\n')

def splitdash (xyz):
	return xyz.split(' - ')

# comments row
sheet.row_dimensions[5].height = 100
sheet['B5'] = 'Comments'
sheet['D5'] = 'Refer to ARM App User guide in step 2 page 19 and update where required'
sheet['E5'] = 'Refer to ARM App User guide in step 2 page 20 and update where required'
sheet['G5'] = 'Refer to ARM App User guide in step 2 page 21 - 23 and update where required.\n\nPlease review the unapproved risks'
sheet['H5'] = 'Refer to ARM App User guide in step 2 page 24 and update where required'
sheet['I5'] = 'Refer to ARM App User guide in step 2 page 25 and update where required'
sheet['K5'] = 'Refer to ARM App User guide in step 2 page 26 - 27 and update where required'
sheet['L5'] = 'Refer to ARM App User guide in step 2 page 28 - 29 and update where required'
sheet['M5'] = 'Refer to ARM App User guide in step 2 page 30 and update'
sheet['N5'] = 'Refer to ARM App User guide in step 2 page 32-33 and update where required'
sheet['O5'] = 'Refer to ARM App User guide in step 2 page 34 and update where required'
sheet['U5'] = 'Refer to ARM App User guide in step 2 page 62 and update.\nReview and update providing a brief concise note of how you arrived at the assessment\n"Probability of X% based on …"\n"Impact of X-Y days to recover"'
sheet['V5'] = 'Refer to ARM App User guide in step 2 page 49 and update.\n\nNote the dates past the resolution date. Have these been resolved? If so has the risk been updated or if not what is happening?'
sheet['W5'] = 'Refer to ARM App User guide in step 2 page 50 and update where required.\n\nPlease  note these dates should be aligned to your schedule activity that you have noted in the activity field where applicable'
sheet['X5'] = 'Refer to ARM App User guide in step 2 page 51 and update where required.\n\nPlease  note these dates should be aligned to your schedule activity that you have noted in the activity field where applicable'
sheet['AD5'] = 'Any high risk to be reviewed and approved?'
sheet['AK5'] = 'Refer to ARM App User guide in step 2 page 79-80 and update where required.\n\nAre there any additional actions?'
sheet['AL5'] = 'Refer to ARM App User guide in step 2 page 81 and update where required.\n\nHave the action owners changed for any outstanding actions?'
sheet['AM5'] = 'Refer to ARM App User guide in step 2 page 82 and update where required.\n\nHave any active actions been completed?'
sheet['AN5'] = 'Refer to ARM App User guide in step 2 page 83 and update where required.'
sheet['AR5'] = 'Are there any overdue actions?\nExclamation marks (!!!) at the end means the action is overdue'
sheet['AS5'] = 'Have all completed actions been updated and if completed then has the current risk rating been re-assessed?'

# list to compare with a cell
UPDATE = ['Undefined', 'Unapproved', '0 : NIL']
OVERDUE = []
COLUMN = {'Folder': int,
          'ID': int,
          'Title': int,
          'Owner': int,
          'Status': int,
          'Cause': int,
          'Effect': int,
          'Current Controls': int,
          'Description': int,
          'WBS': int,
          'Key Element': int,
          'Category': int,
          'Reviewed By': int,
          'Last Review': int,
          'Next Review Date': int,
          'Last Review Note': int,
          'Rationale': int,
          'Target Resolution': int,
          'Trigger Date': int,
          'Expiry Date': int,
          'Current Probability (Qualitative)': int,
          'Current Schedule Impact': int,
          'Current Cost Impact': int,
          'Current Reputation Impact': int,
          'Current HSE Impact': int,
          'Current Risk Level': int,
          'Target Probability (Qualitative)': int,
          'Target Schedule Impact': int,
          'Target Cost Impact': int,
          'Target Reputation Impact': int,
          'Target HSE Impact': int,
          'Target Risk Level': int,
          'Action Title': int,
          'Action Owner': int,
          'Action Status': int,
          'Action Type': int,
          'Action Description': int,
          'Action Comments': int,
          'Action Start Date': int,
          'Due Date': int,
          'Completion Date': int}

# looping the sheet
for rowNum in range(5, sheet.max_row + 1):
    for columnNum in range(1, sheet.max_column + 1):

        # setting up variables for loop
        cell = sheet.cell(row=rowNum, column=columnNum)

        # settign the conditions
        if rowNum == 5 and cell.value is not None:
            cell.font = italic
            cell.fill = yellowFill
            cell.alignment = alignment
            cell.border = border

        # getting column number into variable
        elif rowNum == 6 and cell.value in COLUMN:
            COLUMN[cell.value] = columnNum

        # cell is blank or in UPDATE
        elif rowNum > 6 and ((cell.value == '' or cell.value is None or cell.value in UPDATE) or 
                          (columnNum == COLUMN['Title'] and cell.value.startswith("Delay")) or
                          ((columnNum == COLUMN['Current Cost Impact']
                                  or columnNum == COLUMN['Current Reputation Impact']
                                  or columnNum == COLUMN['Current HSE Impact']
                                  or columnNum == COLUMN['Target Cost Impact']
                                  or columnNum == COLUMN['Target Reputation Impact']
                                  or columnNum == COLUMN['Target HSE Impact']) and cell.value != 'NIL') or
                          (columnNum == COLUMN['Action Owner'] and cell.value.find("Undefined") != -1)
                          ):
            cell.fill = redFill

		# current controls and action title	cross reference
        elif rowNum > 6 and (columnNum == COLUMN['Current Controls'] or columnNum == COLUMN['Action Title']):
            cxreflist = splitline(cell.value)
            cxreflist = [x for x in cxreflist if x != ''] # deleting blank lines
            for idex in cxreflist:
                #if idex.strip().find("(C") == -1 and idex.strip().find("[C") == -1 and idex.strip().find("[c") == -1 and idex.strip().find("(c") == -1 and idex.strip().find("- C") ==-1 and idex.strip().find("( C") ==-1:
                action = idex.replace(" ", "").lower()
                if action.find("(c") == -1 and action.find("[c") == -1:# and idex.strip().find("[c") == -1 and idex.strip().find("(c") == -1 and idex.strip().find("- C") ==-1 and idex.strip().find("( C") ==-1:
                  cell.fill = redFill
                else:
                    continue
        
        # current rating lesser than target rating
        elif rowNum > 6 and columnNum == COLUMN['Current Risk Level']:
            tarrate = sheet.cell(row=rowNum, column=COLUMN['Target Risk Level'])
            curratelist = cell.value.split(' : ')  # split the numbers and word
            tarratelist = tarrate.value.split(' : ')  # split the numbers and word
            if int(curratelist[0]) < int(tarratelist[0]):  # compare the number rating
                cell.fill = redFill

        # actions
        elif rowNum > 6 and columnNum == COLUMN['Action Status']:
            currate = sheet.cell(row=rowNum, column=COLUMN['Current Risk Level'])
            # actions that are completed but rating not re-assessed
            if cell.value.find("Active") == -1 and currate.value != tarrate.value:  # compare current and target rating
                cell.fill = redFill

            # overdue actions
            else:
                #stsplt = cell.value.split('\n')  # getting each line of action status into list
                stslist = splitline(cell.value)
                duedates = sheet.cell(row=rowNum, column=COLUMN['Due Date'])

                for idx, val in enumerate(stslist):
                    if val.endswith("Active"):
                        duedatelist = splitline(duedates.value)#.split('\n')  # getting each line of action duedate into list
                        date_line = splitdash(duedatelist[idx])#.split(' - ')  # splitting the action id and date into list. d[0] is the action id. d[1] is the date to compare.
                        if date_line[1] != '':  # if the date is not empty
                            due = datetime.strptime(date_line[1], '%d %b %y').date()  # converting str type into datetime type
                            if due < today:  # compare with today's date
                                duedatelist[idx] = duedatelist[idx] + " !!!" #this will just update the corresponding date in list
                                OVERDUE.append(idx)
                                #duedates.value = newduedates
                                duedates.value = '\n'.join(duedatelist)
                                duedates.fill = redFill
                        else:  # if date is empty
                            duedates.fill = redFill

sheet['AR4'] = str(len(OVERDUE)) + ' Overdue Action(s)'
sheet.title='Comments'
wb.save('03-comments.xlsx')
